New 'Who's Online' dashboard widget

davidpede opened the pull request:
### What does it do?
- Changed from unstyled table layout to a modx grid
- New processor to return online users
- Improved accuracy of 20min window
- Add 'manager_use_fullname' setting support

### Why is it needed?
Improves design consistency and fixes users showing in list when last activity was +20mins. Also address a security concern where administrators user names are revealed as 'manager_use_fullname' setting is ignored.

### Related issue(s)/PR(s)
#11518

davidpede committed:
New getonline processor

Update processor

Update tpl

Update controller

New grid

Add 'manager_date_format' and 'manager_time_format' setting support

Fix hardcoded datetime zone

Remove date from current time

Add 'manager_use_fullname' setting support

Lexicon load in addHtml

Add pagination to grid

davidpede commented:
Currently `manager_date_format` & `manager_time_format` sys settings don't have an effect on the 'last hit' or 'current time' format but probably should.

Will also look at supporting `manager_use_fullname` for the 'user' column.

I'll try to add tomorrow.

Mark-H commented on manager/controllers/default/dashboard/widget.grid-online.php:
```diff
@@ -7,52 +7,26 @@
+    $this->modx->response->addLangTopic('core:dashboard');
 
-        $users = array();
+    $datetime = new DateTime('Europe/London');
```
Probably need to fix this hardcoded timezone

OptimusCrime commented:
Care to provide a print screen of the widget? Code looks good :)

davidpede commented:
@OptimusCrime Sure, it looks like this:

![chrome_2017-07-11_13-43-38](https://user-images.githubusercontent.com/825106/28068973-a0f60584-663f-11e7-80bd-4edfc5b33653.png)

Mark-H commented:
Thanks @davidpede. 

Looks like the tests are failing due to this PR, any thoughts on that?

````
1) modDashboardTest::testRender
Error: Call to a member function addLangTopic() on null
/home/travis/build/modxcms/revolution/manager/controllers/default/dashboard/widget.grid-online.php:14
/home/travis/build/modxcms/revolution/core/model/modx/moddashboardwidget.class.php:235
/home/travis/build/modxcms/revolution/core/model/modx/moddashboardwidget.class.php:107
/home/travis/build/modxcms/revolution/core/model/modx/moddashboard.class.php:78
/home/travis/build/modxcms/revolution/_build/test/Tests/Model/Dashboard/modDashboardTest.php:69
````

davidpede commented:
I'm looking into the fails but have struggled with where or how to load the `dashboard` lexicon topic from the start. It's line 14 in the widget.grid-online.php file that is causing the problem. However this is the only method out everything I've found online that worked and loaded the topic. Any ideas welcome!

davidpede commented:
Just pushing some changes that hopefully sort it...

davidpede commented:
Sorry but I have hit the limit of my knowledge with this one. The line causing the test fail is: https://github.com/davidpede/revolution/blob/f4f23075e6d00a7cde82e56768a0f7fbde9abf35/manager/controllers/default/dashboard/widget.grid-online.php#L14 (it passes when I comment it out).

I am obviously loading the lexicon topic incorrectly but have seen it done like this in other cmp/extras and it works. Have tried everything I can find but no dice.

Will need assistance to progress this pr.

rtripault commented:
I believe it's possible that `modX::$controller` might not be defined/set at this stage.

Anyway, if you look at the `modDashboardWidgetInterface` constructor, you will notice some controller is available there (and later using `$this->controller`).

I hope that would help!

davidpede commented:
Thanks @rtripault I've tried `$this->controller->addLexiconTopic('core:dashboard');` as well and it also works in my modx dev install but Travis fails it for a undefined `modX::$smarty` property.

Also tried:

```
public function getContent() {
    $this->modx->controller->addLexiconTopic('core:dashboard');
}
```

But it doesn't work in my modx dev install. the lexicon placeholders are blank.

rtripault commented:
I'm afraid without extra commits i won't be able to provide guidance

OptimusCrime commented:
Please try `$this->controller->addLexiconTopic('core:dashboard');`

Ref: https://github.com/modxcms/revolution/blob/9277f3da567724aa2223e70edebd5de080076dd3/core/model/modx/moddashboardwidget.class.php#L229

davidpede commented:
@OptimusCrime have just changed it to `$this->controller->addLexiconTopic('core:dashboard');` and now fails on `Undefined property: modX::$smarty`

rtripault commented:
Yes, Travis is failing maybe because of stricter PHP configuration than your environment.

Basically `addLexiconTopic()` tries to assign variables to Smarty, which might not be available i guess.

Another option might be `$this->modx->lexicon->load()` instead.

Also, you might want to remove your image files (and others not related to modx core files) from your PR.

Let us know how it goes!

Mark-H commented:
Perhaps it's the test that's wrong 😁 

In [MODxControllerTestCase](https://github.com/modxcms/revolution/blob/2.x/_build/test/MODxControllerTestCase.php#L47) there's some additional setup being done to initiate smarty, which is missing from the test that keeps failing in this PR, [modDashboardTest](https://github.com/modxcms/revolution/blob/2.x/_build/test/Tests/Model/Dashboard/modDashboardTest.php). 

As the widget was simpler before, just returning HTML, that test wouldn't have needed to load it specifically in the past, but now that it uses a smarty template it does. 

`$this->controller->addLexiconTopic` is likely the best way to load the lexicon in this case, but this PR could do with a bit of a cleanup, so perhaps worth creating a fresh one when it's passing the test?

davidpede commented:
Sorry last commit had stuff in it shouldn't, have removed.

davidpede commented:
Still stuck on this and starting to think @Mark-H is right about the test haha!

To summarise:

1. `$this->controller->addLexiconTopic('core:dashboard');` fails test because `Undefined property: modX::$smarty`
2. `$this->modx->controller->addLexiconTopic('core:dashboard');` fails test because `Call to a member function addLexiconTopic() on a non-object`
3. `$this->modx->lexicon->load('core:dashboard')` doesn't load lexicon in modx manager
4. `$this->modx->controller->lexicon->load('core:dashboard')` fails in modx manager with error `Fatal error: Call to a member function load() on null`

Most other examples I've found of loading lexicon topic in a controller uses `$this->modx->controller->addLexiconTopic('core:dashboard');`

Out of ideas tbh

davidpede commented:
Have managed to get this working and pass the tests. This is what did it: https://github.com/davidpede/revolution/blob/ca9a72b70450fa5bf5988a6b8d3bce02cf67da8d/manager/controllers/default/dashboard/widget.grid-online.php#L24

Let me know if you'd like a fresh, cleaner pr and I'll create one

OptimusCrime commented:
@davidpede If you squash the commits I think this should be good to go!

davidpede commented:
Cheers @OptimusCrime! I've squashed all the travis check commits, is that enough or you want it as one commit?

ps: also added pagination to the grid for sites with tons of users

OptimusCrime commented:
Squashing commits means turning multiple commits into one commit. This is usually done if you have a large number of commits to make it easier to track the history of changes back in time. Now it would seem like you were the author of the changes of some lines 11 times in a row. 

However, I am not in charge of this. Perhaps the other guys are OK with the commits as they are.

davidpede commented:
Have squashed into one, looks like it worked ok

OptimusCrime commented:
Looks like it did! Awesome work :D

opengeek commented:
Cherry-picked into 2.x as 689ca8bea — please target 2.x in the future.
